internal/mode/static/nginx/config/maps.go

```diff
@@ -61,6 +61,23 @@ func createStreamMaps(conf dataplane.Configuration) []*http.Map {
 		}
 	}
 
+	for _, s := range conf.SSLServers {
+		streamMap, ok := portsToMap[s.Port]
+
+		hostname := s.Hostname
+
+		if s.IsDefault {
+			hostname = "default"
```

Comment:
In some cases we generate more than one default(ish) server block for an HTTPS port.

There are two cases: 
(1) HTTPS listener with a hostname
(2) HTTPS listener with no hostname --> matches all hostnames

In the first case, we generate two server blocks:
- a default server
- and a server for the hostname

In the second case, we generate 2+ server blocks
- a default server
- an ssl server with the wildcard hostname that terminates using the cert/key provided in the listener
- an ssl server block for each route with a distinct hostname


Example of case (1) with one route for cafe.example.com and /tea:

```
server {
    listen 443 ssl default_server;

    ssl_reject_handshake on;
}

server {
    listen 443 ssl;
    ssl_certificate /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;
    ssl_certificate_key /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;

    if ($ssl_server_name != $host) {
        return 421;
    }

    server_name cafe.example.com;

    location = /tea {
        proxy_set_header Host "$gw_api_compliant_host";
        proxy_set_header X-Forwarded-For "$proxy_add_x_forwarded_for";
        proxy_set_header Upgrade "$http_upgrade";
        proxy_set_header Connection "$connection_upgrade";
        proxy_pass http://default_tea_80$request_uri;

        proxy_http_version 1.1;
    }

    location / {
        return 404 "";
    }
}
```

Example for case (2) with the same route:

```
server {
    listen 443 ssl default_server;

    ssl_reject_handshake on;
}

server {
    listen 443 ssl;
    ssl_certificate /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;
    ssl_certificate_key /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;

    if ($ssl_server_name != $host) {
        return 421;
    }

    server_name cafe.example.com;
    location = /tea {

        proxy_set_header Host "$gw_api_compliant_host";
        proxy_set_header X-Forwarded-For "$proxy_add_x_forwarded_for";
        proxy_set_header Upgrade "$http_upgrade";
        proxy_set_header Connection "$connection_upgrade";
        proxy_pass http://default_tea_80$request_uri;

        proxy_http_version 1.1;
    }

    location / {
        return 404 "";

    }

}


server {
    listen 443 ssl;
    ssl_certificate /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;
    ssl_certificate_key /etc/nginx/secrets/ssl_keypair_certificate_cafe-secret.pem;

    if ($ssl_server_name != $host) {
        return 421;
    }

    server_name ~^;
    location / {
        return 404 "";
    }
}

```

`server_name ~^` matches all server names.

We need to make sure we generate the correct config in both these cases. 

TBH I'm not exactly sure how to do that right now, I'll need to look into a bit more and do some testing. @pleshakov tagging you in case you have some input here.

Reply:
I think that'll work @pleshakov.

Just placing the documentation on the `hostnames` keyword here so we can refer to it later:

hostnames
indicates that source values can be hostnames with a prefix or suffix mask:

```
*.example.com 1;
example.*     1;
```
The following two records

```
example.com   1;
*.example.com 1;
```
can be combined:

```
.example.com  1;
```

This parameter should be specified before the list of values.

Reply:
I am confused about this. Suppose the user created a gateway tls listener to listen on *.example.com port 443 with passthrough enabled and an attached tls route on app.example.com. What would be the expected behavior for an https request on port 443 to cafe.example.com and an https request on port 443 to app.com?

Reply:
Because TLS listener is for TCP connections over TLS, we cannot assume the underlying protocol is HTTP. And 404 response only applies for HTTP protocol.
Connections to ` cafe.example.com` and `app.com` in this case should get a TLS error from NGINX in this case. Which can be configure using a snippet like this:
```
server {
    listen <insert port or unix socket here...> ssl;

    ssl_reject_handshake on;
}
```

Reply:
When I try to add a server like that I get this error:

```
nginx: [emerg] no handler for server in /opt/homebrew/etc/nginx/nginx.conf:53
```

Talked to @kate-osborn and it seems this directive only works when terminating TLS connection.

Reply:
curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to app2.example.com:8444

this is the current error that we get

Reply:
> Talked to @kate-osborn and it seems this directive only works when terminating TLS connection

if you put that server in http context, it should work.

If you put it in the stream block, we need to add a handler - meaning do something with the connection. even though we won't do anything because it will be rejected.

you can try to add `return "";` to that server.

Reply:
> you can try to add return ""; to that server.

Yep, funny, we just got the same advice from the core team. Tested it and it worked.